### PR TITLE
feat(bigtable): add support for request stats

### PR DESCRIPTION
**Pull request by humbertowastaken**
feat(bigtable): add support for request stats

### CODE REVIEW

File: bigtable/bigtable.go
Comment (mutianf) [inline bigtable/bigtable.go:482]: Hmm I'm unfamiliar with go syntax but why do we need to getRequestStatsCallback for rowFilter or Row limit options? And what does getRequestStatsCallback do?

**Review by mutianf — COMMENTED** [review]
Is there a way to add a test for this?

File: bigtable/bigtable.go
Comment (alexoneill) [reply bigtable/bigtable.go:482]: Since the `ReadOption` interface includes the new method `getRequestStatsCallback` signature, for `rowFilter` and `limitRows` to adhere to this interface, they need to include an implementation of that method.

Go's interface system is not totally clear to readers IMO, but Go will automatically mark `rowFilter` and `limitRows` as implementing this interface at compile time as long as the methods match up.

Without these changes, `RowFilter` above breaks since it indicates that it returns a `ReadOption`, which means that `rowFilter` must have a method that matches the `getRequestStatsCallback` spec, else it can't be considered as part of this interface.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:465]: Please do not expose raw proto types in the manual client layers.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:470]: This is very wordy for Go. I would shorten this to `RequestStatsFunc`.

Comment (telpirion) [inline bigtable/bigtable.go:470]: If I read this correctly, every distinct type that implements the ReadOption interface must now have the `getRequestStatsCallback()` function. As this is a private method, it won't cause a breaking change in the API surface, but it will require refactoring all existing ReadOption types.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:498]: This is also a bit wordy. I would shorten to `WithRequestStats`. The type of the parameter tells the developer that it's a function.

**Review by telpirion — CHANGES_REQUESTED** [review]

**Comment by telpirion** [comment]
This PR also requires a unit test and an integration test.

File: bigtable/bigtable.go
Comment (humbertowastaken) [reply bigtable/bigtable.go:498]: Fixed.

**Comment by humbertowastaken** [comment]
Added integration tests.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:466]: Need comments for fields in the `ReadIterationStats` struct.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:476]: Need comments for fields in the `RequestLatencyStats` struct.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:481]: Need comments for fields in the `FullReadStats` struct.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:486]: NIT: a small comment before this function will be useful for future Golang client maintainers.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:487]: Although this is guarded by an if statement above, I recommend a nil check here.

File: bigtable/bigtable.go
Comment (telpirion) [inline bigtable/bigtable.go:491]: NIT: Preferred style for Go is to instantiate an object by providing explicit field assignments.

E.g.

```
return FullReadStats {
    CellsReturnedCount: readStats.CellsReturnedCount,
    // etc etc
}
```

**Review by telpirion — CHANGES_REQUESTED** [review]
I think this still needs a unit test in `bigtable_test.go`. The Go monorepo doesn't run integration tests for each PR until it has been merged. As a result, we need to ensure that unit tests exist to help us see breakages before a PR is merged.

In sum, bare minimum required changes:

+ unit test(s) for this callback feature in `bigtable_test.go`
+ comments for all publicly exposed fields to populate the reference docs

File: bigtable/bigtable.go
Comment (humbertowastaken) [reply bigtable/bigtable.go]: Done.

File: bigtable/bigtable.go
Comment (humbertowastaken) [reply bigtable/bigtable.go:466]: Done.

File: bigtable/bigtable.go
Comment (igorbernstein2) [reply bigtable/bigtable.go:507]: What would you recommend to do in the nil case? Since this is an internal impl detail, we can update this in a follow up PR

**Comment by igorbernstein2** [comment]
At this point the only thing that missing is a unit test. This will require some refactoring in the emulator code, which we can do in a follow up PR. The functionality is covered by the integration test. Given the timing, I'm going to go ahead and merge this.

**Review by igorbernstein2 — APPROVED** [review]
lgtm.

**Pull request merged**
